mettle/src/coreapi.c

```diff
@@ -125,6 +115,97 @@ static struct tlv_packet *core_uuid(struct tlv_handler_ctx *ctx)
 static struct tlv_packet *core_negotiate_tlv_encryption(struct tlv_handler_ctx *ctx)
 {
 	return tlv_packet_response_result(ctx, TLV_RESULT_SUCCESS);
+} 
+
+static struct tlv_packet *core_loadlib(struct tlv_handler_ctx *ctx)
+{
+	uint32_t flags;
+	size_t extension_len;
+	struct mettle *m = ctx->arg;
+	struct tlv_dispatcher *td = mettle_get_tlv_dispatcher(m);
+	struct tlv_packet *p = tlv_packet_response(ctx);
+	int fd = -1;
+	char *full_path = NULL;
+	int tlv_result = TLV_RESULT_FAILURE;
+	const char *library_path = tlv_packet_get_str(ctx->req, TLV_TYPE_LIBRARY_PATH);
+	const unsigned char *extension = tlv_packet_get_raw(ctx->req, TLV_TYPE_DATA, &extension_len);
+
+	tlv_packet_get_u32(ctx->req, TLV_TYPE_FLAGS, &flags);
+	if (!extension || extension_len == 0) {
+		log_error("No extension received");
+		goto done;
+	}
+
+	if (!library_path) {
+		log_error("No extension name specified");
+		goto done;
+	}
+
+	/* Check if this is a binary image or executable. */
+	const char bin_magic_number[] = BIN_MAGIC_NUMBER;
+	if (strncmp((const char *)&extension[extension_len - sizeof(bin_magic_number)], bin_magic_number, sizeof(bin_magic_number)) == 0) {
+		log_info("Loading extension '%s' from binary image", library_path);
+
+		// Make a copy of this image, don't use the TLV buffer...
+		unsigned char *extension_copy = malloc(extension_len);
+		if (extension_copy == NULL) {
+			log_error("Failed to allocate memory for '%s' binary image", library_path);
+			goto done;
+		}
+		// TODO: free this mem when the extension no longer is running
+		memcpy(extension_copy, extension, extension_len);
+		if (extension_start_binary_image(m, library_path, extension, extension_len, NULL))
+		{
+			log_error("Failed to start extension from binary image '%s'", library_path);
+			goto done;
+		}
+	} else {
+		/* This is an executable that needs to be saved to the filesystem. */
+		log_info("Loading extension '%s' from executable file", library_path);
+
+		if (asprintf(&full_path, "/tmp/%s", library_path) <=0) {
```

Comment:
I think we should replace `full_path` here with 
`const char *target_path = tlv_packet_get_str(ctx->req, TLV_TYPE_TARGET_PATH);` ?
Then the framework api can send `/tmp/ext(random)` or `/data/local/tmp/ext(random)` on Android?
Currently this is failing on Android aarch64 because /tmp doesn't exist.

Reply:
Man, yeah, definitely...!  😅